### PR TITLE
AB#39005: Generic Repo Replace

### PR DESCRIPTION
**Pull request by MoizIqbal97**
PR which gets rid of GenericEFRepository
Removes some of the unused Read and Write service methods

### CODE REVIEW

**Review by dcl10 — APPROVED** [review]

**Comment by AndyRae** [comment]
It's worth deleting the actual GenericRepository class + interface as well since it's no longer in use

File: src/Submissions/Api/Services/Directory/BiobankWriteService.cs [hunk 1]
```diff
@@ -154,55 +95,61 @@ public async Task UpdateOrganisationAnnualStatisticAsync(int organisationId, int
...
+            await _context.SaveChangesAsync();
         }
 
         public async Task AddBiobankRegistrationReasons(List<OrganisationRegistrationReason> activeRegistrationReasons)
         {
             foreach (var registrationReason in activeRegistrationReasons)
             {
                 //Validate reason id first - don't want to go around inserting new unnamed reasons
-                if (await _registrationReasonRepository.GetByIdAsync(registrationReason.RegistrationReasonId) != null)
+                if (await _context.RegistrationReasons.FindAsync(registrationReason.RegistrationReasonId) != null)
```
Comment (beforan) [inline src/Submissions/Api/Services/Directory/BiobankWriteService.cs:122]:
```suggestion
                if (await _db.RegistrationReasons.FindAsync(registrationReason.RegistrationReasonId) != null)
```

File: src/Submissions/Api/Services/Directory/BiobankWriteService.cs [hunk 2]
```diff
@@ -154,55 +95,61 @@ public async Task UpdateOrganisationAnnualStatisticAsync(int organisationId, int
...
                 {
                     //now make sure the biobank doesn't already have this reason listed
-                    if ((await _organisationRegistrationReasonRepository.ListAsync(false,
-                        x => x.OrganisationId == registrationReason.OrganisationId && x.RegistrationReasonId == registrationReason.RegistrationReasonId))
-                        .FirstOrDefault() == null)
+                    var biobank = await _context.OrganisationRegistrationReasons
```
Comment (beforan) [inline src/Submissions/Api/Services/Directory/BiobankWriteService.cs:125]:
```suggestion
                    var biobank = await _db.OrganisationRegistrationReasons
```

File: src/Submissions/Api/Services/Directory/BiobankWriteService.cs [hunk 3]
```diff
@@ -154,55 +95,61 @@ public async Task UpdateOrganisationAnnualStatisticAsync(int organisationId, int
...
+                    .Where(x => x.OrganisationId == registrationReason.OrganisationId && x.RegistrationReasonId == registrationReason.RegistrationReasonId)
+                    .ToListAsync();
+
+                    if (biobank.FirstOrDefault() == null) 
                     {
-                        _organisationRegistrationReasonRepository.Insert(registrationReason);
+                        _context.OrganisationRegistrationReasons.Add(registrationReason);
```
Comment (beforan) [inline src/Submissions/Api/Services/Directory/BiobankWriteService.cs:131]:
```suggestion
                        _db.OrganisationRegistrationReasons.Add(registrationReason);
```

File: src/Submissions/Api/Services/Directory/BiobankWriteService.cs [hunk 4]
```diff
@@ -154,55 +95,61 @@ public async Task UpdateOrganisationAnnualStatisticAsync(int organisationId, int
...
                     }
                 }
             }
-            await _organisationRegistrationReasonRepository.SaveChangesAsync();
+            await _context.SaveChangesAsync();
```
Comment (beforan) [inline src/Submissions/Api/Services/Directory/BiobankWriteService.cs:135]:
```suggestion
            await _db.SaveChangesAsync();
```

File: src/Submissions/Api/Services/Directory/BiobankWriteService.cs [hunk 5]
```diff
@@ -154,55 +95,61 @@ public async Task UpdateOrganisationAnnualStatisticAsync(int organisationId, int
...
         }
 
         public async Task DeleteBiobankRegistrationReasonAsync(int organisationId, int registrationReasonId)
         {
             //make sure the biobank has this reason
-            if ((await _organisationRegistrationReasonRepository.ListAsync(false,
-                x => x.OrganisationId == organisationId && x.RegistrationReasonId == registrationReasonId))
-                .FirstOrDefault() != null)
+            var biobank = await _context.OrganisationRegistrationReasons
```
Comment (beforan) [inline src/Submissions/Api/Services/Directory/BiobankWriteService.cs:141]:
```suggestion
            var biobank = await _db.OrganisationRegistrationReasons
```

File: src/Submissions/Api/Services/Directory/BiobankWriteService.cs [hunk 6]
```diff
@@ -154,55 +95,61 @@ public async Task UpdateOrganisationAnnualStatisticAsync(int organisationId, int
...
+            .Where(x => x.OrganisationId == organisationId && x.RegistrationReasonId == registrationReasonId)
+            .ToListAsync();
+
+            var entity = _context.OrganisationRegistrationReasons.Where(x => x.OrganisationId == organisationId && x.RegistrationReasonId == registrationReasonId).FirstOrDefault();
```
Comment (beforan) [inline src/Submissions/Api/Services/Directory/BiobankWriteService.cs:145]:
```suggestion
            var entity = _db.OrganisationRegistrationReasons.Where(x => x.OrganisationId == organisationId && x.RegistrationReasonId == registrationReasonId).FirstOrDefault();
```

File: src/Submissions/Api/Services/Directory/BiobankWriteService.cs [hunk 7]
```diff
@@ -154,55 +95,61 @@ public async Task UpdateOrganisationAnnualStatisticAsync(int organisationId, int
...
+
+          if (biobank.FirstOrDefault() != null) 
             {
-                await
-                _organisationRegistrationReasonRepository.DeleteWhereAsync(
-                    x => x.OrganisationId == organisationId && x.RegistrationReasonId == registrationReasonId);
-            }
+             _context.OrganisationRegistrationReasons.Remove(entity);
```
Comment (beforan) [inline src/Submissions/Api/Services/Directory/BiobankWriteService.cs:149]:
```suggestion
             _db.OrganisationRegistrationReasons.Remove(entity);
```

File: src/Submissions/Api/Services/Directory/BiobankWriteService.cs [hunk 8]
```diff
@@ -154,55 +95,61 @@ public async Task UpdateOrganisationAnnualStatisticAsync(int organisationId, int
...
+            };
 
-            await _organisationRegistrationReasonRepository.SaveChangesAsync();
+            await _context.SaveChangesAsync();
```
Comment (beforan) [inline src/Submissions/Api/Services/Directory/BiobankWriteService.cs:152]:
```suggestion
            await _db.SaveChangesAsync();
```

File: src/Submissions/Api/Services/Directory/DiseaseStatusService.cs [hunk 9]
```diff
@@ -1,21 +1,27 @@
...
+using Microsoft.EntityFrameworkCore;
 using System.Linq;
 using System.Threading.Tasks;
 
 namespace Biobanks.Submissions.Api.Services.Directory
 {
     public class DiseaseStatusService : IDiseaseStatusService
     {
-        private readonly IGenericEFRepository<SnomedTag> _snomedTagRepository;
+    private readonly ApplicationDbContext _context;
```
Comment (beforan) [inline src/Submissions/Api/Services/Directory/DiseaseStatusService.cs:12]:
```suggestion
    private readonly ApplicationDbContext _db;
```

File: src/Submissions/Api/Services/Directory/DiseaseStatusService.cs [hunk 10]
```diff
@@ -1,21 +1,27 @@
...
 
-        public DiseaseStatusService (
-            IGenericEFRepository<SnomedTag> snomedTagRepository
+    public DiseaseStatusService (
+            ApplicationDbContext context
```
Comment (beforan) [inline src/Submissions/Api/Services/Directory/DiseaseStatusService.cs:15]:
```suggestion
            ApplicationDbContext db
```

File: src/Submissions/Api/Services/Directory/DiseaseStatusService.cs [hunk 11]
```diff
@@ -1,21 +1,27 @@
...
             )
         {
-            _snomedTagRepository = snomedTagRepository;
-        }
-        public async Task<SnomedTag> GetSnomedTagByDescription(string description)
-            => (await _snomedTagRepository.ListAsync(filter: x => x.Value == description)).SingleOrDefault();
+      _context = context;
```
Comment (beforan) [inline src/Submissions/Api/Services/Directory/DiseaseStatusService.cs:18]:
```suggestion
      _db = db;
```

File: src/Submissions/Api/Services/Directory/DiseaseStatusService.cs [hunk 12]
```diff
@@ -1,21 +1,27 @@
...
     }
+    public async Task<SnomedTag> GetSnomedTagByDescription(string description)
+       => await _context.SnomedTags
```
Comment (beforan) [inline src/Submissions/Api/Services/Directory/DiseaseStatusService.cs:21]:
```suggestion
       => await _db.SnomedTags
```

File: src/Submissions/Api/Services/Directory/SqlServerLogoStorageProvider.cs [hunk 13]
```diff
@@ -9,23 +11,21 @@ namespace Biobanks.Submissions.Api.Services.Directory
 {
     public class SqlServerLogoStorageProvider : ILogoStorageProvider
     {
-        private readonly IGenericEFRepository<Blob> _blobRepository;
-        private readonly IGenericEFRepository<Organisation> _organisationRepository;
+      private readonly ApplicationDbContext _context;
```
Comment (beforan) [inline src/Submissions/Api/Services/Directory/SqlServerLogoStorageProvider.cs:14]:
```suggestion
      private readonly ApplicationDbContext _db;
```

File: src/Submissions/Api/Services/Directory/SqlServerLogoStorageProvider.cs [hunk 14]
```diff
@@ -9,23 +11,21 @@ namespace Biobanks.Submissions.Api.Services.Directory
...
 
-        public SqlServerLogoStorageProvider(
-            IGenericEFRepository<Blob> blobRepository,
-            IGenericEFRepository<Organisation> organisationRepository)
+      public SqlServerLogoStorageProvider(
+          ApplicationDbContext context)
```
Comment (beforan) [inline src/Submissions/Api/Services/Directory/SqlServerLogoStorageProvider.cs:17]:
```suggestion
          ApplicationDbContext db)
```

File: src/Submissions/Api/Services/Directory/SqlServerLogoStorageProvider.cs [hunk 15]
```diff
@@ -9,23 +11,21 @@ namespace Biobanks.Submissions.Api.Services.Directory
...
         {
-            _blobRepository = blobRepository;
-            _organisationRepository = organisationRepository;
+          _context = context;
```
Comment (beforan) [inline src/Submissions/Api/Services/Directory/SqlServerLogoStorageProvider.cs:19]:
```suggestion
          _db = db;
```

File: src/Submissions/Api/Services/Directory/SqlServerLogoStorageProvider.cs [hunk 16]
```diff
@@ -9,23 +11,21 @@ namespace Biobanks.Submissions.Api.Services.Directory
...
         }
 
         public async Task<Blob> GetLogoBlobAsync(string resourceName)
         {
-            var blob = (await _blobRepository.ListAsync(
-                false,
-                x => x.FileName == resourceName))
-                .FirstOrDefault();
+        
+            var blob = await _context.Blobs
```
Comment (beforan) [inline src/Submissions/Api/Services/Directory/SqlServerLogoStorageProvider.cs:25]:
```suggestion
            var blob = await _db.Blobs
```

File: src/Submissions/Api/Services/Directory/SqlServerLogoStorageProvider.cs [hunk 17]
```diff
@@ -48,26 +48,25 @@ public async Task<string> StoreLogoAsync(MemoryStream logo, string fileName, str
...
             //we want to replace, not keep adding and storing new files
-            try
-            {
-                var existing = await GetLogoBlobAsync(logoBlob.FileName);
-                await _blobRepository.DeleteAsync(existing.Id);
-            }
-            catch (ApplicationException) { } //no worries if nothing found
+     
+            var existing = await GetLogoBlobAsync(logoBlob.FileName);
+              _context.Remove(existing.Id);
```
Comment (beforan) [inline src/Submissions/Api/Services/Directory/SqlServerLogoStorageProvider.cs:53]:
```suggestion
              _db.Remove(existing.Id);
```

File: src/Submissions/Api/Services/Directory/SqlServerLogoStorageProvider.cs [hunk 18]
```diff
@@ -48,26 +48,25 @@ public async Task<string> StoreLogoAsync(MemoryStream logo, string fileName, str
...
 
             //write to db
-            _blobRepository.Insert(logoBlob);
-            await _blobRepository.SaveChangesAsync();
+            _context.Add(logoBlob);
```
Comment (beforan) [inline src/Submissions/Api/Services/Directory/SqlServerLogoStorageProvider.cs:56]:
```suggestion
            _db.Add(logoBlob);
```

File: src/Submissions/Api/Services/Directory/SqlServerLogoStorageProvider.cs [hunk 19]
```diff
@@ -48,26 +48,25 @@ public async Task<string> StoreLogoAsync(MemoryStream logo, string fileName, str
...
+            await _context.SaveChangesAsync();
```
Comment (beforan) [inline src/Submissions/Api/Services/Directory/SqlServerLogoStorageProvider.cs:57]:
```suggestion
            await _db.SaveChangesAsync();
```

File: src/Submissions/Api/Services/Directory/SqlServerLogoStorageProvider.cs [hunk 20]
```diff
@@ -48,26 +48,25 @@ public async Task<string> StoreLogoAsync(MemoryStream logo, string fileName, str
...
 
             return logoBlob.FileName;
         }
 
         public async Task RemoveLogoAsync(int organisationId)
         {
-            var organisation = await _organisationRepository.GetByIdAsync(organisationId);
-
+            var organisation = await _context.FindAsync<Organisation>(organisationId);
```
Comment (beforan) [inline src/Submissions/Api/Services/Directory/SqlServerLogoStorageProvider.cs:64]:
```suggestion
            var organisation = await _db.FindAsync<Organisation>(organisationId);
```

File: src/Submissions/Api/Services/Directory/SqlServerLogoStorageProvider.cs [hunk 21]
```diff
@@ -48,26 +48,25 @@ public async Task<string> StoreLogoAsync(MemoryStream logo, string fileName, str
...
+            var entity = _context.Blobs.Where(x => x.FileName == organisation.Logo);
```
Comment (beforan) [inline src/Submissions/Api/Services/Directory/SqlServerLogoStorageProvider.cs:65]:
```suggestion
            var entity = _db.Blobs.Where(x => x.FileName == organisation.Logo);
```

File: src/Submissions/Api/Services/Directory/SqlServerLogoStorageProvider.cs [hunk 22]
```diff
@@ -48,26 +48,25 @@ public async Task<string> StoreLogoAsync(MemoryStream logo, string fileName, str
...
             if (organisation != null)
-                await _blobRepository.DeleteWhereAsync(x => x.FileName == organisation.Logo);
-        }
+                _context.Remove(entity);
```
Comment (beforan) [inline src/Submissions/Api/Services/Directory/SqlServerLogoStorageProvider.cs:67]:
```suggestion
                _db.Remove(entity);
```

File: src/Submissions/Api/Services/Directory/SqlServerLogoStorageProvider.cs [hunk 23]
```diff
@@ -48,26 +48,25 @@ public async Task<string> StoreLogoAsync(MemoryStream logo, string fileName, str
...
+             
+            await _context.SaveChangesAsync();
```
Comment (beforan) [inline src/Submissions/Api/Services/Directory/SqlServerLogoStorageProvider.cs:69]:
```suggestion
            await _db.SaveChangesAsync();
```

**Review by dcl10 — APPROVED** [review]

**Review by beforan — APPROVED** [review]

**Pull request merged**